### PR TITLE
Port to Windows (Microsoft Visual Studio 2015)

### PR DESCRIPTION
**Pull request by davidalbertonogueira**
Necessary modifications for StarSpace compilation in Windows with Microsoft Visual Studio 2015.

### CODE REVIEW

**Comment by ledw** [comment]
@davidalbertonogueira Thanks for your work on this! I will review your pull request soon.

File: src/doc_data.cpp [hunk 1]
```diff
@@ -39,20 +40,20 @@ void LayerDataHandler::loadFromFile(
   foreach_line(
     fileName,
     [&](std::string& line) {
-      auto& corpus = corpora[getThreadID()];
-      ParseResults example;
-      if (parser->parse(line, example)) {
-        corpus.push_back(example);
-      }
-    },
+    auto& corpus = corpora[getThreadID()];
```
Comment (ledw) [inline src/doc_data.cpp]: Why do we change indentation here?

File: src/doc_data.cpp [hunk 2]
```diff
@@ -126,33 +140,33 @@ void LayerDataHandler::convert(
...
     } else
-    if (args_->trainMode == 2) {
-      // pick one random rhs as lhs, the rest becomes rhs features
-      auto idx = rand() % example.RHSFeatures.size();
-      for (int i = 0; i < example.RHSFeatures.size(); i++) {
-        if (i == idx) {
-          insert(rslt.LHSTokens, example.RHSFeatures[i], args_->dropoutLHS);
-        } else {
-          insert(rslt.RHSTokens, example.RHSFeatures[i], args_->dropoutRHS);
+      if (args_->trainMode == 2) {
```
Comment (ledw) [inline src/doc_data.cpp]: I tink we don't need to change indentation here as it's consistent across files.

File: src/model.cpp [hunk 3]
```diff
@@ -169,8 +204,8 @@ Real EmbedModel::trainOneExample(
 
 Real EmbedModel::train(shared_ptr<InternDataHandler> data,
                        int numThreads,
-		       std::chrono::time_point<std::chrono::high_resolution_clock> t_start,
-		       int epochs_done,
+           std::chrono::time_point<std::chrono::high_resolution_clock> t_start,
```
Comment (ledw) [inline src/model.cpp]: indentation here too

**Comment by davidalbertonogueira** [comment]
The changes in indentation weren't intentional (it was an involuntary Visual Studio 'format document').
I will fix it.

**Comment by NAZMULBABLU** [comment]
Ok we can restore  all

**Comment by ledw** [comment]
@davidalbertonogueira thanks! Yes please fix the indentation and update the PR. :-)

**Comment by jwijffels** [comment]
would love to see this included so that we can proceed on #118 !

File: src/doc_data.cpp (hunk 1 above)
Comment (davidalbertonogueira) [reply src/doc_data.cpp]: Fixed

File: src/doc_data.cpp (hunk 2 above)
Comment (davidalbertonogueira) [reply src/doc_data.cpp]: fixed

File: src/model.cpp (hunk 3 above)
Comment (davidalbertonogueira) [reply src/model.cpp]: fixed

**Comment by davidalbertonogueira** [comment]
Involuntary Visual Studio 'format document' reverted.

**Comment by ledw** [comment]
@davidalbertonogueira Please make sure all the test cases pass and the example scripts under examples/ folder works the same before commit.

**Comment by davidalbertonogueira** [comment]
@ledw , the executables of the format *_test (eg. matrix_test, proj_test) pass all their tests in my Windows environment.
However, the bash scripts under examples/ folder are, per definition, Linux ones. Those scripts as they are, they cannot be run in Windows. 

Someone would need to create some equivalent Windows batch files for that. I would recommend closing and accepting this request as it is (port concerning the code), and opening a new issue & pull request for the community to start working on that.

**Comment by jwijffels** [comment]
This really looks promising, this at least compiles on my Windows machine when building an R wrapper around it. Haven't looked at any bugs but at least it compiles. Thank you!

**Comment by davidalbertonogueira** [comment]
>

**Comment by davidalbertonogueira** [comment]
Yes, @ledw, it does. I've just checked it: compiling and successfully running the tests under Ubuntu 16.04 xenial.

**Comment by ledw** [comment]
@davidalbertonogueira Great. Thanks!

**Pull request merged**